File: src/utils/vscode.ts
// NodeJS
import fs from 'node:fs/promises'
import process from 'node:process'

// Utils
import { writeMessage } from './console'
import { getErrorMessage } from './errors'
import { createEmptyJsonFile, exists } from './user-os'

// Constants
import { UTF8_ENCODING } from '@/constants/encoding'

// Types
import type { Extensions, Settings } from '@/types/vscode'

export async function configureVSCode() {
  try {
    writeMessage({
      type: 'info',
      message: 'Configuring VSCode...'
    })

    const existVSCodeSettingsFile = await exists(
      `${process.cwd()}/.vscode/settings.json`
    )

    if (!existVSCodeSettingsFile) {
      await createEmptyJsonFile('.vscode/settings.json')
    }

    const vscodeSettings = await fs.readFile('.vscode/settings.json', {
      encoding: UTF8_ENCODING
    })

    const vscodeSettingsObject: Settings = JSON.parse(vscodeSettings)

    vscodeSettingsObject['[css]'] = JSON.stringify({
      'editor.defaultFormatter': 'stylelint.vscode-stylelint'
    })

    await fs.writeFile(
      '.vscode/settings.json',
      JSON.stringify(vscodeSettingsObject, null, 2),
      {
        encoding: UTF8_ENCODING
      }
    )

    writeMessage({
      type: 'info',
      message: 'VSCode configured successfully'
    })
  } catch {
    writeMessage({
      type: 'error',
      message: getErrorMessage('VSCodeConfig')
    })

    process.exit(1)
  }
}

export async function addRecommendedExtension(extension: string) {
  try {
    writeMessage({
      type: 'info',
      message: 'Adding recommended extensions...'
    })

    const existVSCodeSettingsFile = await exists(
      `${process.cwd()}/.vscode/extensions.json`
    )

    if (!existVSCodeSettingsFile) {
      await createEmptyJsonFile('.vscode/extensions.json')
    }

    const vscodeExtensions = await fs.readFile('.vscode/extensions.json', {
      encoding: UTF8_ENCODING
    })

    const vscodeExtensionsObject: Extensions = JSON.parse(vscodeExtensions)

    if (vscodeExtensionsObject.recommendations === undefined) {
      vscodeExtensionsObject.recommendations = []

      await fs.writeFile(
        '.vscode/extensions.json',
        JSON.stringify(vscodeExtensionsObject, null, 2),
        {
          encoding: UTF8_ENCODING
        }
      )
    }

    vscodeExtensionsObject.recommendations.push(extension)

    await fs.writeFile(
      '.vscode/extensions.json',
      JSON.stringify(vscodeExtensionsObject, null, 2),
      {
        encoding: UTF8_ENCODING
      }
    )

    writeMessage({
      type: 'info',
      message: 'Recommended extensions added successfully'
    })
  } catch {
    writeMessage({
      type: 'error',
      message: getErrorMessage('AddRecommendedExtensions')
    })

    process.exit(1)
  }
}

File: src/utils/vscode.go
package utils

import (
	"encoding/json"
	"os"
	"stylelintbc/src/types"
)

func ConfigureVSCode() {
	WriteMessage(WriteMessageProps{
		Type:    "info",
		Message: "Configuring VSCode...",
	})

	existVSCodeFolder := Exists(".vscode")

	if !existVSCodeFolder {
		CreateFolder(".vscode")
	}

	existVSCodeSettingsFile := Exists(".vscode/settings.json")

	if !existVSCodeSettingsFile {
		CreateEmptyJsonFile(".vscode/settings")
	}

	data, err := os.ReadFile(".vscode/settings.json")

	if err != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("VSCodeConfig")),
		})

		os.Exit(1)
	}

	var settings types.VSCodeSettings

	deserializeErr := json.Unmarshal(data, &settings)

	if deserializeErr != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("VSCodeConfig")),
		})

		os.Exit(1)
	}

	settings["[css]"] = map[string]interface{}{
		"editor.defaultFormatter": "stylelint.vscode-stylelint",
	}

	jsonData, serializeErr := json.MarshalIndent(settings, "", "  ")

	if serializeErr != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("VSCodeConfig")),
		})

		os.Exit(1)
	}

	writeFileErr := os.WriteFile(".vscode/settings.json", jsonData, 0644)

	if writeFileErr != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("VSCodeConfig")),
		})

		os.Exit(1)
	}

	WriteMessage(WriteMessageProps{
		Type:    "success",
		Message: "VSCode configured successfully",
	})
}

func AddRecommendedExtension(extension string) {
	WriteMessage(WriteMessageProps{
		Type:    "info",
		Message: "Adding recommended extensions...",
	})

	existVSCodeFolder := Exists(".vscode")

	if !existVSCodeFolder {
		CreateFolder(".vscode")
	}

	existVSCodeExtensionsFile := Exists(".vscode/extensions.json")

	if !existVSCodeExtensionsFile {
		CreateEmptyJsonFile(".vscode/extensions")
	}

	data, err := os.ReadFile(".vscode/extensions.json")

	if err != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("AddRecommendedExtensions")),
		})

		os.Exit(1)
	}

	var extensions types.VSCodeExtensions

	deserializeErr := json.Unmarshal(data, &extensions)

	if deserializeErr != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("AddRecommendedExtensions")),
		})

		os.Exit(1)
	}

	if extensions.Recommendations == nil {
		extensions.Recommendations = []string{}
	}

	extensions.Recommendations = append(extensions.Recommendations, extension)

	jsonData, serializeErr := json.MarshalIndent(extensions, "", "  ")

	if serializeErr != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("AddRecommendedExtensions")),
		})

		os.Exit(1)
	}

	writeFileErr := os.WriteFile(".vscode/extensions.json", jsonData, 0644)

	if writeFileErr != nil {
		WriteMessage(WriteMessageProps{
			Type:    "error",
			Message: string(GetErrorMessage("AddRecommendedExtensions")),
		})

		os.Exit(1)
	}

	WriteMessage(WriteMessageProps{
		Type:    "success",
		Message: "Recommended extensions added successfully",
	})
}
